File: android/app/src/main/java/com/rafaeltmbr/stopwatch/core/services/Stopwatch.kt
package com.rafaeltmbr.stopwatch.core.services

import com.rafaeltmbr.stopwatch.core.utils.ObservableState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

enum class StopwatchStatus {
    INITIAL,
    RUNNING,
    PAUSED
}

data class StopwatchState(
    val seconds: Float = 0f,
    val status: StopwatchStatus = StopwatchStatus.INITIAL
)

class Stopwatch {
    private var _state: ObservableState<StopwatchState> = ObservableState(StopwatchState())

    val state: ObservableState<StopwatchState>
        get() = this._state

    private var job: Job? = null

    fun start() {
        if (_state.state.status != StopwatchStatus.INITIAL) return

        loop()
    }

    fun resume() {
        if (_state.state.status != StopwatchStatus.PAUSED) return

        loop()
    }

    fun pause() {
        if (_state.state.status != StopwatchStatus.RUNNING) return

        job?.cancel()
        job = null
        _state.updateState(_state.state.copy(status = StopwatchStatus.PAUSED))
    }

    fun reset() {
        if (_state.state.status != StopwatchStatus.PAUSED) return

        _state.updateState(StopwatchState())
    }

    private fun loop() {
        _state.updateState(_state.state.copy(status = StopwatchStatus.RUNNING))

        job = CoroutineScope(Dispatchers.Default).launch {
            while (_state.state.status == StopwatchStatus.RUNNING) {
                _state.updateState(
                    _state.state.copy(seconds = _state.state.seconds + 0.1f)
                )

                delay(100L)
            }
        }
    }
}

File: ios/Stopwatch/core/services/Stopwatch.swift
import Foundation

enum StopwatchStatus {
    case INITIAL
    case RUNNING
    case PAUSED
}

struct StopwatchState: Equatable, Hashable {
    var seconds: Float = 0.0
    var status: StopwatchStatus = .INITIAL
}

class Stopwatch<S: StateObserver<StopwatchState>> {
    private(set) var state = ObservableState<S>(StopwatchState())
    
    private var task: Task<(), Never>? = nil

    func start() {
        guard state.state.status == .INITIAL else { return }
        
        loop()
    }
    
    func resume() {
        guard state.state.status == .PAUSED else { return }
        
        loop()
    }
    
    func pause() {
        guard state.state.status == .RUNNING else { return }
        
        task?.cancel()
        task = nil
        state.updateState(StopwatchState(seconds: state.state.seconds, status: .PAUSED))
    }
    
    func reset() {
        guard state.state.status == .PAUSED else { return }
        
        state.updateState(StopwatchState())
    }
    
    private func loop() {
        state.updateState(StopwatchState(seconds: state.state.seconds, status: .RUNNING))
        
        task = Task {@MainActor in
            while state.state.status == .RUNNING {
                state.updateState(StopwatchState(seconds: state.state.seconds + 0.1, status: .RUNNING))
                try? await Task.sleep(for: .milliseconds(100))
            }
        }
    }
}
